### PR TITLE
Let search phases override max concurrent requests

### LINKED ISSUE #26198: Dockerized Elasticsearch instance crashes when receiving request
**Elasticsearch version** (`bin/elasticsearch --version`):

Version: 6.0.0-beta1, Build: 896afa4/2017-08-03T23:14:26.258Z, JVM: 1.8.0_144

**Plugins installed**: []

**JVM version** (`java -version`):

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):

Linux 816fd3d99829 4.4.0-91-generic #114-Ubuntu SMP Tue Aug 8 11:56:56 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

I'm running Elasticsearch within a docker container (Docker version 17.05.0-ce, build 89658be). Have used older versions up to Elasticsearch 5.5.1 with the same setup without any problems.

When testing to run Elasticsearch 6.0.0-beta1 I have faced an issue that I haven't seem before. The issue is that Elasticsearch crashes when receiving a request.

When starting up the Elasticsearch container, it starts up and Elasticsearch health status seems ok by looking in the logs as it states `Cluster health status changed from [RED] to [YELLOW]` and only a single Elasticsearch instance is running.

Running a simple API  towards the elasticsearch instance works fine:

```
user@host:/elasticsearch# curl localhost:9200/
{
  "name" : "dvphuFi",
  "cluster_name" : "elasticsearch-logs",
  "cluster_uuid" : "k8QHY2SRTsyZn8sLo5vNRw",
  "version" : {
    "number" : "6.0.0-beta1",
    "build_hash" : "896afa4",
    "build_date" : "2017-08-03T23:14:26.258Z",
    "build_snapshot" : false,
    "lucene_version" : "7.0.0",
    "minimum_wire_compatibility_version" : "5.6.0",
    "minimum_index_compatibility_version" : "5.0.0"
  },
  "tagline" : "You Know, for Search"
}
```

However, running any query will cause the Elasticsearch instance to crash:

```
user@host:/elasticsearch# curl localhost:9200/_search?q=test
curl: (52) Empty reply from server
```

The following steps was done in order to get the system working again after the error started happening:
1. Stop, remove and restart the docker image -> Error still occurred
2. Restarting the Docker deamon (via `service docker restart` -> Error still occurred
3. Restarting the Linux host -> Error still occurred
4. Deleting the Elasticsearch data dir -> System works ok again, however data needs to be re-indexed

More logs will be attached as a comment to this issue.

Below is an excerpt from the logs:

```
[2017-08-14T11:20:27,152][INFO ][o.e.n.Node               ] [] initializing ...
[2017-08-14T11:20:27,257][INFO ][o.e.e.NodeEnvironment    ] [dvphuFi] using [1] data paths, mounts [[/data (/dev/sda1)]], net usable_space [14.5gb], net total_space [31.3gb], types [ext4]
[2017-08-14T11:20:27,257][INFO ][o.e.e.NodeEnvironment    ] [dvphuFi] heap size [3.8gb], compressed ordinary object pointers [true]
[2017-08-14T11:20:27,591][INFO ][o.e.n.Node               ] node name [dvphuFi] derived from node ID [dvphuFiRT8aEJicgCiXprg]; set [node.name] to override
[2017-08-14T11:20:27,591][INFO ][o.e.n.Node               ] version[6.0.0-beta1], pid[14], build[896afa4/2017-08-03T23:14:26.258Z], OS[Linux/4.4.0-91-generic/amd64], JVM[Oracle Corporation/Java HotSpot(TM) 64-Bit Server VM/1.8.0_144/25.144-b01]
[2017-08-14T11:20:27,592][INFO ][o.e.n.Node               ] JVM arguments [-Xms4g, -Xmx4g, -Dlog4j2.disable.jmx=true, -Des.path.home=/elasticsearch, -Des.path.conf=/elasticsearch/config]
[2017-08-14T11:20:27,592][WARN ][o.e.n.Node               ] version [6.0.0-beta1] is a pre-release version of Elasticsearch and is not suitable for production
[2017-08-14T11:20:28,200][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [aggs-matrix-stats]
[2017-08-14T11:20:28,200][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [analysis-common]
[2017-08-14T11:20:28,201][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [ingest-common]
[2017-08-14T11:20:28,201][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [lang-expression]
[2017-08-14T11:20:28,201][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [lang-mustache]
[2017-08-14T11:20:28,202][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [lang-painless]
[2017-08-14T11:20:28,202][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [parent-join]
[2017-08-14T11:20:28,202][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [percolator]
[2017-08-14T11:20:28,202][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [reindex]
[2017-08-14T11:20:28,202][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [repository-url]
[2017-08-14T11:20:28,202][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [transport-netty4]
[2017-08-14T11:20:28,203][INFO ][o.e.p.PluginsService     ] [dvphuFi] loaded module [tribe]
[2017-08-14T11:20:28,203][INFO ][o.e.p.PluginsService     ] [dvphuFi] no plugins loaded
[2017-08-14T11:20:29,526][INFO ][o.e.d.DiscoveryModule    ] [dvphuFi] using discovery type [zen]
[2017-08-14T11:20:30,304][INFO ][o.e.n.Node               ] initialized
[2017-08-14T11:20:30,305][INFO ][o.e.n.Node               ] [dvphuFi] starting ...
[2017-08-14T11:20:30,336][INFO ][i.n.u.i.PlatformDependent] Your platform does not provide complete low-level API for accessing direct buffers reliably. Unless explicitly requested, heap buffer will always be preferred to avoid potential system instability.
[2017-08-14T11:20:30,431][INFO ][o.e.t.TransportService   ] [dvphuFi] publish_address {172.18.0.10:9300}, bound_addresses {0.0.0.0:9300}
[2017-08-14T11:20:30,440][INFO ][o.e.b.BootstrapChecks    ] [dvphuFi] bound or publishing to a non-loopback or non-link-local address, enforcing bootstrap checks
[2017-08-14T11:20:33,486][INFO ][o.e.c.s.MasterService    ] [dvphuFi] zen-disco-elected-as-master ([0] nodes joined), reason: new_master {dvphuFi}{dvphuFiRT8aEJicgCiXprg}{riLrClxxSlugm8Ks7PMClg}{172.18.0.10}{172.18.0.10:9300}
[2017-08-14T11:20:33,492][INFO ][o.e.c.s.ClusterApplierService] [dvphuFi] new_master {dvphuFi}{dvphuFiRT8aEJicgCiXprg}{riLrClxxSlugm8Ks7PMClg}{172.18.0.10}{172.18.0.10:9300}, reason: apply cluster state (from master [master {dvphuFi}{dvphuFiRT8aEJicgCiXprg}{riLrClxxSlugm8Ks7PMClg}{172.18.0.10}{172.18.0.10:9300} committed version [1] source [zen-disco-elected-as-master ([0] nodes joined)]])
[2017-08-14T11:20:33,526][INFO ][o.e.h.n.Netty4HttpServerTransport] [dvphuFi] publish_address {172.18.0.10:9200}, bound_addresses {0.0.0.0:9200}
[2017-08-14T11:20:33,526][INFO ][o.e.n.Node               ] [dvphuFi] started
[2017-08-14T11:20:35,029][INFO ][o.e.g.GatewayService     ] [dvphuFi] recovered [125] indices into cluster_state
[2017-08-14T11:20:45,558][INFO ][o.e.c.r.a.AllocationService] [dvphuFi] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[logstash-logs-2017.08.09][2], [logstash-logs-2017.08.09][3]] ...]).
[2017-08-14T11:31:18,825][ERROR][o.e.t.n.Netty4Utils      ] fatal error on the network layer
        at org.elasticsearch.transport.netty4.Netty4Utils.maybeDie(Netty4Utils.java:179)
        at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.exceptionCaught(Netty4HttpRequestHandler.java:81)
        at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
        at io.netty.channel.AbstractChannelHandlerContext.notifyHandlerException(AbstractChannelHandlerContext.java:850)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:364)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at org.elasticsearch.http.netty4.pipelining.HttpPipeliningHandler.channelRead(HttpPipeliningHandler.java:63)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:544)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at java.lang.Thread.run(Thread.java:748)
[2017-08-14T11:31:18,839][WARN ][o.e.h.n.Netty4HttpServerTransport] [dvphuFi] caught exception while handling client http traffic, closing connection [id: 0x423a8b89, L:/127.0.0.1:9200 - R:/127.0.0.1:44298]
java.lang.StackOverflowError: null
        at com.carrotsearch.hppc.ObjectObjectHashMap.<init>(ObjectObjectHashMap.java:123) ~[hppc-0.7.1.jar:?]
...
...
...
```

Full log is attached below.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Start and run Elasticsearch within a Docker container
 2. Run a query against the Elasticsearch instance such as `/_search?q=test`
 3. Elasticsearch crashes with a stacktrace in the logs (see below)

**Provide logs (if relevant)**:

Logs will be added as a comment.

### PR DESCRIPTION
**Pull request by jasontedor**
If the query coordinating node is also a data node that holds all the shards for a search request, we can end up recursing through the can match phase (because we send a local request and on response in the listener move to the next shard and do this again, without ever having returned from previous shards). This recursion can lead to stack overflow for even a reasonable number of indices (daily indices over a sixty days with five shards per day is enough to trigger the stack overflow). Moreover, all this execution would be happening on a network thread (the thread that initially received the query). With this commit, we allow search phases to override max concurrent requests. This allows the can match phase to avoid recursing through the shards towards a stack overflow.

Closes #26198

### CODE REVIEW

**Comment by jasontedor** [comment]
This is fine in 5.6.0, the request are forked to the search thread pool there already.

File: core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+public class CanMatchIT extends ESIntegTestCase {
+
+    /*
+     * Previously the can match phase executed on the same thread pool. In cases that a query coordinating node held all the shards for a
+     * query, the can match phase would recurse and end in stack overflow (and this thread would be a networking thread, tying such a thread
+     * up for an extended period of time). This test is for that situation.
+     */
+    public void testAvoidStackOverflow() throws InterruptedException {
+        final String node = internalCluster().startDataOnlyNode(Settings.builder().put("node.attr.color", "blue").build());
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java]: why do we need to start a node? can't we use an existing node and use it's _name for allocation filtering?

File: core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.elasticsearch.test.ESIntegTestCase;
+
+public class CanMatchIT extends ESIntegTestCase {
+
+    /*
+     * Previously the can match phase executed on the same thread pool. In cases that a query coordinating node held all the shards for a
+     * query, the can match phase would recurse and end in stack overflow (and this thread would be a networking thread, tying such a thread
+     * up for an extended period of time). This test is for that situation.
+     */
+    public void testAvoidStackOverflow() throws InterruptedException {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java]: nit - this isn't really a can match test but more an "can we run with lot's of shards on nodes" tests. I think it's good to have but maybe we can fold it into one of the generic search IT suites? maybe `SimpleSearchIT` (although nothing is simple in life ;)

**Review by bleskes — APPROVED** [review]
LGTM

File: core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        final String node = internalCluster().startDataOnlyNode(Settings.builder().put("node.attr.color", "blue").build());
+        /*
+         * While 640 shards for a single index is excessive, creating one index with 640 shards is cheaper than creating 128 indices with 5
+         * shards yet both set up the failing reproduction that we need.
+         */
+        final Settings.Builder settings =
+                Settings.builder().put("index.routing.allocation.include.color", "blue").put("index.number_of_shards", 640);
+        client().admin().indices().create(new CreateIndexRequest("index").settings(settings)).actionGet();
+        // it can take a long time for all the shards to allocate and initialize
+        ensureGreen(TimeValue.timeValueSeconds(Long.MAX_VALUE));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java]: this is worrisome...  how long does this test run normally? should we fallback to mocking?

**Review by s1monw — CHANGES_REQUESTED** [review]
I think we should fix this differently. This is only an issue since we are running into the `maxConcurrentShardRequests` which is unnecessary for this `can_match` action I wonder if we should rather opt out of it entirely otherwise we are subject to rejections which was partially the purpose of the `can_match` change.

**Comment by jasontedor** [comment]
@s1monw For this reason when I initially submitted the pull request I forked this requests to the generic thread pool. After seeing that in the 5.6 branch these requests are forked to the search thread pool I followed your lead there. I'll look into opting out of max concurrent shard requests and I'll reach out later this week to discuss all of our options.

**Comment by s1monw** [comment]
> @s1monw For this reason when I initially submitted the pull request I forked this requests to the generic thread pool. After seeing that in the 5.6 branch these requests are forked to the search thread pool I followed your lead there. 

lemme provide some history here since it's not obvious, in 5.6 do still fetch resouces in the rewrite phase for some queries. I fixed this in #25791 which allows to execute the rewrite phase on the network thread. The execution on the search threadpool will again cause potential rejections which we tried to prevent with moving to the network threadpool. 

> I'll look into opting out of max concurrent shard requests and I'll reach out later this week to discuss all of our options.

Looking forward to your suggestions

**Comment by jasontedor** [comment]
@bleskes @s1monw I have updated this pull request, would you please take a look?

File: core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java (hunk 3 above)
Comment (jasontedor) [reply core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java]: This test is removed, I am now using mocking.

File: core/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java [hunk 4]
```diff
@@ -52,7 +52,12 @@
     private final AtomicInteger shardExecutionIndex = new AtomicInteger(0);
     private final int maxConcurrentShardRequests;
 
-    InitialSearchPhase(String name, SearchRequest request, GroupShardsIterator<SearchShardIterator> shardsIts, Logger logger) {
+    InitialSearchPhase(
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java]: can we maybe not warp on every parameter please?

File: core/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java [hunk 5]
```diff
@@ -57,7 +53,7 @@
                                         SearchTask task, Function<GroupShardsIterator<SearchShardIterator>, SearchPhase> phaseFactory) {
         super("can_match", logger, searchTransportService, nodeIdToConnection, aliasFilter, concreteIndexBoosts, executor, request,
             listener,
-            shardsIts, timeProvider, clusterStateVersion, task, new BitSetSearchPhaseResults(shardsIts.size()));
+            shardsIts, timeProvider, clusterStateVersion, task, new BitSetSearchPhaseResults(shardsIts.size()), shardsIts.size());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java]: can we leave a comment why we do this here?

**Review by s1monw — APPROVED** [review]
LGTM left 2 minors thanks for doing this

**Review by bleskes — COMMENTED** [review]
Much cleaner. Thanks.

**Pull request merged**